File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.cpp
class Solution {
public:
    long long mod = 1e9+7;
    int memo(vector<vector<int>>& a,int i,int j,int sum,int k,vector<vector<vector<int>>>& dp){
        if(i<0 || j<0)
            return 0;
        if(i==0 && j==0){
            sum+=a[0][0];
            return sum%k==0;
        }
        
        if(dp[i][j][sum%k]!=-1)
            return dp[i][j][sum%k];
        
        int left = memo(a,i,j-1,(sum+a[i][j])%k,k,dp)%mod;
        int right = memo(a,i-1,j,(sum+a[i][j])%k,k,dp)%mod;
        return dp[i][j][sum%k] = (left+right)%mod;
        
    }
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        
        int m = size(grid),n=size(grid[0]);
        vector<vector<vector<int>>> dp(m,vector<vector<int>>(n,vector<int>(51,-1)));

        return memo(grid,m-1,n-1,0,k,dp);
    }
};

File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.java
class Solution {
    int mod = 1000_000_007;
    public int numberOfPaths(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        int[][][] dp = new int[m][n][k];
        for(int[][] arr:dp){
            for(int[] a: arr) 
                Arrays.fill(a, -1);
        }
        
        
        
        return memo(grid, m-1, n-1, 0, k, dp);
    }
    
    int memo(int[][] grid, int m, int n, int sum, int k, int[][][] dp){
        
        if (m<0 || n< 0) return 0;
        
        if (m == 0 && n == 0){
            sum += grid[m][n];
            return sum%k==0?1:0;
        }
        
        if (dp[m][n][sum%k] != -1) return dp[m][n][sum%k];
        
        int left = memo(grid, m, n-1, (sum + grid[m][n])%k, k, dp);
        int right = memo(grid, m-1, n, (sum + grid[m][n])%k, k, dp);
        
        return dp[m][n][sum%k] = (left+right)%mod;
    }
}